cmd/snap-confine/snap-confine.c

```diff
@@ -126,8 +150,106 @@ static void sc_preserve_and_sanitize_process_state(sc_preserved_process_state *
 {
+	/* Restore original umask */
 	umask(proc_state->orig_umask);
 	debug("umask restored to %#4o", proc_state->orig_umask);
+
+	/* Restore original current working directory.
+	 *
+	 * This part is more involved for the following reasons. While we hold an
+	 * O_PATH file descriptor that still points to the original working
+	 * directory, that directory may not be representable in the target mount
+	 * namespace. A quick example may be /custom that exists on the host but
+	 * not in the base snap of the application.
+	 *
+	 * Also consider when the path of the original working directory now
+	 * maps to a different inode we cannot use fchdir(2). One example of
+	 * that is the /tmp directory, which exists in both the host mount
+	 * namespace and the per-snap mount namespace but actually represents a
+	 * different directory.
+	 **/
+
+	/* Read the target of symlink at /proc/self/fd/<fd-of-orig-cwd> */
+	char fd_path[PATH_MAX];
+	char orig_cwd[PATH_MAX];
+	ssize_t nread;
+	/* If the original working directory cannot be used for whatever reason then
+	 * move the process to a special void directory. */
+	const char *sc_void_dir = "/var/lib/snapd/void";
+
+	sc_must_snprintf(fd_path, sizeof fd_path, "/proc/self/fd/%d",
+			 proc_state->orig_cwd_fd);
+	nread = readlink(fd_path, orig_cwd, sizeof orig_cwd);
+	if (nread < 0) {
+		die("cannot read symbolic link target %s", fd_path);
+	}
+	if (nread == sizeof orig_cwd) {
+		die("cannot fit symbolic link target %s", fd_path);
+	}
+
+	/* Open path corresponding to the original working directory in the
+	 * execution environment. This may normally fail if the path no longer
+	 * exists here, this is not a fatal error. */
+	int inner_cwd_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	inner_cwd_fd =
+	    open(orig_cwd, O_PATH | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+	if (inner_cwd_fd < 0 && errno != ENOENT) {
+		die("cannot open path of the original working directory %s (%d)", orig_cwd, errno);
+	}
+	if (inner_cwd_fd < 0) {
+		debug("cannot represent original working directory %s", orig_cwd);
+		goto the_void;
+	}
+
+	/* The original working directory exists in the execution environment
+	 * which lets us check if it points to the same inode as before. */
+	struct stat file_info_inner;
+	if (fstat(inner_cwd_fd, &file_info_inner) < 0) {
+		die("cannot stat path of working directory in the execution environment");
+	}
+
+	/* Note that we cannot use proc_state->orig_cwd_fd as that points to the
+	 * directory but in another mount namespace and using that causes
+	 * weird and undesired effects.
+	 *
+	 * By the time this code runs we are already running as the
+	 * designated user so UNIX permissions are in effect. */
+	if (fchdir(inner_cwd_fd) < 0) {
+		if (errno == EPERM || errno == EACCES) {
+			debug("cannot access original working directory %s", orig_cwd);
+			goto the_void;
+		}
+		die("cannot restore original working directory via path");
+	}
+	if (proc_state->file_info_orig_cwd.st_dev ==
```

Comment:
Maybe add a TODO comment on why we are differentiating since it's just for debugging?

Reply:
Done now.